airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlCdcTargetPosition.java

```diff
@@ -0,0 +1,104 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.integrations.source.mssql;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.base.Preconditions;
+import io.airbyte.db.jdbc.JdbcDatabase;
+import io.airbyte.db.jdbc.JdbcUtils;
+import io.airbyte.integrations.debezium.CdcTargetPosition;
+import io.airbyte.integrations.debezium.internals.SnapshotMetadata;
+import io.debezium.connector.sqlserver.Lsn;
+import java.io.IOException;
+import java.sql.SQLException;
+import java.util.List;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MssqlCdcTargetPosition implements CdcTargetPosition {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MssqlCdcTargetPosition.class);
+  private final Lsn targetLsn;
+
+  public MssqlCdcTargetPosition(Lsn targetLsn) {
+    this.targetLsn = targetLsn;
+  }
+
+  @Override
+  public boolean reachedTargetPosition(JsonNode valueAsJson) {
+    Lsn recordLsn = extractLsn(valueAsJson);
+
+    if (targetLsn.compareTo(recordLsn) > 0) {
+      return false;
+    } else {
+      SnapshotMetadata snapshotMetadata = SnapshotMetadata.valueOf(valueAsJson.get("source").get("snapshot").asText().toUpperCase());
+      // if not snapshot or is snapshot but last record in snapshot.
+      return SnapshotMetadata.TRUE != snapshotMetadata;
+    }
+  }
+
+  private Lsn extractLsn(JsonNode valueAsJson) {
+    return Optional.ofNullable(valueAsJson.get("source"))
+        .flatMap(source -> Optional.ofNullable(source.get("commit_lsn").asText()))
+        .map(Lsn::valueOf)
+        .orElseThrow(() -> new IllegalStateException("Could not find LSN"));
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    MssqlCdcTargetPosition that = (MssqlCdcTargetPosition) o;
+    return targetLsn.equals(that.targetLsn);
+  }
+
+  @Override
+  public int hashCode() {
+    return targetLsn.hashCode();
+  }
+
+  public static MssqlCdcTargetPosition getTargetPostion(JdbcDatabase database) {
+    try {
+      final List<JsonNode> jsonNodes = database
+          .bufferedResultSetQuery(conn -> conn.createStatement().executeQuery("SELECT sys.fn_cdc_get_max_lsn() AS max_lsn;"), JdbcUtils::rowToJson);
+      Preconditions.checkState(jsonNodes.size() == 1);
+      if (jsonNodes.get(0).get("max_lsn") != null) {
+        Lsn maxLsn = Lsn.valueOf(jsonNodes.get(0).get("max_lsn").binaryValue());
+        LOGGER.info("identified target lsn: " + maxLsn);
+        return new MssqlCdcTargetPosition(maxLsn);
+      } else {
+        throw new RuntimeException("Max LSN is null, see docs"); // todo: make this error way better
```

Comment:
what action is required from the user here? can we put it in the error message

Reply:
Addressed in https://github.com/airbytehq/airbyte/pull/4689/commits/2f87d62186ad5bd99b496afe4a0529b9b0e099cb